src/Validation/ScannableURLProvider.php

```diff
@@ -0,0 +1,305 @@
+			foreach ( $public_post_types as $post_type ) {
+				$post_ids = $this->get_posts_that_support_amp( $this->get_posts_by_type( $post_type, $i, 1 ) );
+				if ( ! empty( $post_ids[0] ) ) {
+					$urls[] = [
+						'url'  => get_permalink( $post_ids[0] ),
+						'type' => $post_type,
+					];
+				}
+			}
+
+			foreach ( $amp_enabled_taxonomies as $taxonomy ) {
+				$taxonomy_links = $this->get_taxonomy_links( $taxonomy, $i, 1 );
+				$link           = reset( $taxonomy_links );
+				if ( ! empty( $link ) ) {
+					$urls[] = [
+						'url'  => $link,
+						'type' => $taxonomy,
+					];
+				}
+			}
+
+			$author_page_urls = $this->get_author_page_urls( $i, 1 );
+			if ( ! empty( $author_page_urls[0] ) ) {
+				$urls[] = [
+					'url'  => $author_page_urls[0],
+					'type' => 'author',
+				];
+			}
+		}
+
+		// Only validate 1 date and 1 search page.
+		$url = $this->get_date_page();
+		if ( $url ) {
+			$urls[] = [
+				'url'  => $url,
+				'type' => 'date',
+			];
+		}
+		$url = $this->get_search_page();
+		if ( $url ) {
+			$urls[] = [
+				'url'  => $url,
+				'type' => 'search',
+			];
+		}
+
+		return $urls;
+	}
+
+	/**
+	 * Gets whether the template is supported.
+	 *
+	 * @param string $template The template to check.
+	 * @return bool Whether the template is supported.
+	 */
+	private function is_template_supported( $template ) {
+		// If we received an allowlist of conditionals, this template conditional must be present in it.
+		if ( ! empty( $this->include_conditionals ) ) {
+			return in_array( $template, $this->include_conditionals, true );
+		}
+		if ( $this->include_unsupported ) {
+			return true;
+		}
+
+		$supportable_templates = AMP_Theme_Support::get_supportable_templates();
+
+		// Check whether this taxonomy's template is supported, including in the 'AMP Settings' > 'Supported Templates' UI.
+		return ! empty( $supportable_templates[ $template ]['supported'] );
+	}
+
+	/**
+	 * Gets the post IDs that support AMP.
+	 *
+	 * By default, this only gets the post IDs if they support AMP.
+	 * This means that 'Posts' isn't deselected in 'AMP Settings' > 'Supported Templates'
+	 * and 'Enable AMP' isn't unchecked in the post's editor.
+	 *
+	 * @param int[] $ids The post IDs to check for AMP support.
+	 * @return array The post IDs that support AMP, or an empty array.
+	 */
+	private function get_posts_that_support_amp( $ids ) {
+		if ( ! $this->is_template_supported( 'is_singular' ) ) {
+			return [];
+		}
+
+		if ( $this->include_unsupported ) {
+			return $ids;
+		}
+
+		return array_filter(
+			$ids,
+			'post_supports_amp'
+		);
+	}
+
+	/**
+	 * Gets the IDs of public, published posts.
+	 *
+	 * @param string   $post_type The post type.
+	 * @param int|null $offset The offset of the query (optional).
+	 * @param int|null $number The number of posts to query for (optional).
+	 * @return int[]   $post_ids The post IDs in an array.
+	 */
+	private function get_posts_by_type( $post_type, $offset = null, $number = null ) {
```

Comment:
Any reason why the `$number` parameter for `get_posts_by_type()`, `get_posts_by_type()` and `get_taxonomy_links()` have different different default values? I'd recommend standardizing them, if possible.